### PR TITLE
Issue #1383 Add support for ``ALLOCATION_OPTION.stage_to_riv_bot_drn_above`` in ``River.from_imod5_data``.

### LINKED ISSUE #1383: LHM conversion: Rivers removed where stage equals bottom_elevation
The LHM features quite some river cells where stage equals bottom_elevation, effectively making them drains. Especially in the tertiary river system. iMOD Python, in its allocation scheme, now removes these, whereas iMOD5 doesn't.

### PR DESCRIPTION
**Pull request by JoerivanEngelen**
Fixes #1383 and #1060

It turned out the differences in #1383 were mainly caused by a lack of support of ``ALLOCATION_OPTION.stage_to_riv_bot_drn_above``, which is the default setting for iMOD5, but we didn't support it yet.
This was because we lacked some logic to do something with the extra addition Drainage package created by with this option during allocation. This PR implements the following:

- Support for ``ALLOCATION_OPTION.stage_to_riv_bot_drn_above`` in ``River.from_imod5_data``.
- Set ``ALLOCATION_OPTION.stage_to_riv_bot_drn_above`` as default allocation option for rivers
- Adapt some unittests to run with the new default option
- Add two unittests cases to ensure rivers are properly allocated and cleaned when ``stage`` equals ``bottom_elevation``. I first thought there was a bug in one of these functions, so I added this edge case. But they ran without issue! Still good to keep these tests for this edge case to ensure no mistakes creep into the code base over time.

# Checklist
- [x] Links to correct issue
- [x] Update changelog, if changes affect users
- [x] PR title starts with ``Issue #nr``, e.g. ``Issue #737``
- [x] Unit tests were added
- [ ] **If feature added**: Added/extended example

### CODE REVIEW

File: imod/mf6/riv.py [hunk 1]
```diff
@@ -37,11 +37,65 @@
...
+    enforce_dim_order,
+    has_negative_layer,
+    is_planar_grid,
+)
 from imod.util.expand_repetitions import expand_repetitions
 
 
+def maybe_rise_bottom_elevation(bottom_elevation, bottom):
+    """
+    Due to regridding, the bottom_elevation could be smaller than the
```
Comment (Manangka) [inline imod/mf6/riv.py:52]: smaller => less

File: imod/mf6/riv.py [hunk 2]
```diff
@@ -37,11 +37,65 @@
...
+    layer bottom, so here we overwrite it with bottom if that's
+    the case.
+    """
+    layer_bottom_above_bottom_elevation = bottom > bottom_elevation
```
Comment (Manangka) [inline imod/mf6/riv.py:56]: I would split this method up and have the check outside of this method.
The name of this method can then just be `rise_bottom_elevation`

On the calling side it would become:
```
is_layer_above_bottom_elevation = (bottom > bottom_elevation).any()
if is_layer_above_bottom_elevation :
  rise_bottom_elevation(bottom_elevation, bottom)
  logging.logger.warning(message="Note: riv bottom was detected below model bottom. Updated the riv's bottom." )

```

Note:
You can call the warning method from the logger, That wat you dont' have to specify the loglevel manually,

**Review by Manangka — APPROVED** [review]

File: imod/mf6/riv.py [hunk 3]
```diff
@@ -384,22 +445,25 @@ def from_imod5_data(
 
         infiltration_factor = regridded_package_data.pop("infiltration_factor")
 
+        allocation_drn_data: GridDataDict = {}
```
Comment (Manangka) [inline imod/mf6/riv.py:448]: As discussed, this method is rather large and does multiple complicated things.
It's better to split it up to improve the readability

File: imod/mf6/riv.py (hunk 1 above)
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:52]: Done

File: imod/mf6/riv.py (hunk 2 above)
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:56]: Thanks for the instructions on the logger, that reduces code. I didn't split the method into two separate method, as it already is quite small.

File: imod/mf6/riv.py (hunk 3 above)
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:448]: I cleaned the method up, it improved readability a lot!

File: imod/mf6/riv.py [hunk 4]
```diff
@@ -384,22 +445,25 @@ def from_imod5_data(
...
         if is_planar:
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:449]: Done!

**Review by Manangka — APPROVED** [review]
Nice improvement!

File: imod/mf6/riv.py [hunk 5]
```diff
@@ -37,11 +37,113 @@
...
+) -> None:
+    """Set repeat stress for optional package if repeat is not None."""
+    if repeat is not None:
+        if optional_package is not None:
+            optional_package.set_repeat_stress(
+                expand_repetitions(repeat, time_min, time_max)
+            )
+
+
+def maybe_create_package(package: BoundaryCondition) -> Optional[BoundaryCondition]:
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:64]: Renamed to ``mask_package__drop_if_empty``

File: imod/mf6/riv.py [hunk 6]
```diff
@@ -37,11 +37,113 @@
...
+    """ "
+    Create an optional package from a package if it has data. Return None if
+    package is inactive everywhere.
+    """
+    optional_package: Optional[BoundaryCondition] = None
+    # remove River package if its mask is False everywhere
+    mask = ~np.isnan(package["conductance"])
+    if np.any(mask):
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:72]: Done

File: imod/mf6/riv.py [hunk 7]
```diff
@@ -37,11 +37,113 @@
...
+        optional_package = package.mask(mask)
+    return optional_package
+
+
+def maybe_rise_bottom_elevation(bottom_elevation, bottom):
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:77]: Fixed

File: imod/mf6/riv.py [hunk 8]
```diff
@@ -37,11 +37,113 @@
...
+    # create a drainage package with the conductance we computed from the
+    # infiltration factor
+    drn_pkg_data = {
+        "elevation": riv_pkg_data["stage"],
+        "conductance": drain_conductance,
+    }
+    return riv_pkg_data, drn_pkg_data
+
+
+def create_drain_from_leftover_riv_imod5_data(
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:118]: Yes, fixed

File: imod/mf6/riv.py [hunk 9]
```diff
@@ -37,11 +37,113 @@
...
+    """
+    Create a drainage package from leftover imod5 river package data,
+    stemming from:
+
+        * If ``ALLOCATION_OPTION.stage_to_riv_bottom_drn_above`` is chosen,
+            drain cells are allocated from the first active cell to river
+            stage.
+        * Infiltration factor. This factor is optional in imod5, but it
+            does not exist in MF6, so we mimic its effect with a Drainage
+            boundary.
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:131]: Done

File: imod/mf6/riv.py [hunk 10]
```diff
@@ -248,62 +350,69 @@ def allocate_and_distribute_planar_data(
...
             planar_data["stage"],
             planar_data["bottom_elevation"],
         )
-
-        layered_data = {}
-        layered_data["conductance"] = distribute_riv_conductance(
+        # Distribution of conductances
+        allocated_for_distribution = (
+            riv_allocated | drn_allocated
+            if (drn_allocated is not None)
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:365]: Done

File: imod/mf6/riv.py [hunk 11]
```diff
@@ -355,176 +464,66 @@ def from_imod5_data(
...
+        data = {varname: imod5_data[key][varname] for varname in varnames}
+        # Regrid the input data
+        regridded_riv_pkg_data = regrid_imod5_pkg_data(
+            River, data, target_dis, regridder_types, regrid_cache
         )
-
-        infiltration_factor = regridded_package_data.pop("infiltration_factor")
-
-        if is_planar:
+        # Pop infiltration_factor to avoid unnecessisarily allocating and
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/riv.py:481]: Fixed
